tests/TestCase/Controller/Traits/Integration/LoginTraitIntegrationTest.php

```diff
@@ -240,4 +240,48 @@ public function testLogoutNoUser()
         $this->get('/logout');
         $this->assertRedirect('/login');
     }
+
+    /**
+     * Test redirect should not happen if the host is not defined as a known host
+     *
+     * @return void
+     */
+    public function testRedirectAfterLoginToHostUnknown()
+    {
+        $this->post('/login?redirect=http://unknown.com/', [
+            'username' => 'user-4',
+            'password' => '12345',
+        ]);
+        $this->assertRedirect('/pages/home');
+    }
+
+    /**
+     * Test redirect should happen for defaul localhost
+     *
+     * @return void
+     */
+    public function testRedirectAfterLoginToAllowedHost()
+    {
+        Configure::write('Users.AllowedRedirectHosts', ['example.com']);
+        $this->post('/login?redirect=http://example.com/login', [
+            'username' => 'user-4',
+            'password' => '12345',
+        ]);
+        // /login is authorized for this user, and example.com is in the allowed hosts
+        $this->assertRedirect('http://example.com/login');
+    }
+
+    /**
+     * Test redirect fails if url is not allowed
+     *
+     * @return void
+     */
+    public function testRedirectFailsIfUrlNotAllowed()
+    {
+        $this->post('/login?redirect=http://localhost/not-allowed', [
+            'username' => 'user-4',
+            'password' => '12345',
+        ]);
+        $this->assertRedirect('/pages/home');
+    }
```

Comment:
Would be nice to add a test that cover redirect querystring has the full url from current server example:
`https://my.website.com/login?redirect=https://my.website.com/categories`